build_runner/lib/src/generate/build.dart

```diff
@@ -106,17 +106,19 @@ Future<BuildResult> build(List<BuilderApplication> builders,
   );
   var terminator = Terminator(terminateEventStream);
   try {
-    var build = await BuildRunner.create(
-      options,
-      environment,
-      builders,
-      builderConfigOverrides,
-      isReleaseBuild: isReleaseBuild,
-    );
-    var result =
-        await build.run({}, buildDirs: buildDirs, buildFilters: buildFilters);
-    await build.beforeExit();
-    return result;
+    return await runWithFileSystemBatch(() async {
```

Comment:
I think I would prefer to add a new method to `RunnerAssetWriter` which gets invoked at the end of each build. Then we can avoid the zones and just use normal asset reader/writer overriding.

It would be breaking but I think worth it, breaking changes to `build_runner_core` are not really very painful.